README.MD
SwashRL
=======

Branch Directory
----------------

This readme file is used to explain the branch structure for the SwashRL git
repository.  The SwashRL project currently has public repositories at the
following locations:

 * [GitHub](https://github.com/swashdev/SwashRL)  
   Because GitHub is the central hub for the SwashRL project, all links in
   the branch descriptions will link to GitHub pages.

 * [repo.or.cz](https://repo.or.cz/SwashRL.git)

### Public Branches

What follows is a list of branches which currently exist in the repository and
a brief description of their function.  Branches are listed in order of
proximity to the `master` branch; however, note that
[`hotfix-*` and `release-*` branches](#Production-Branches) will be the
closest.

 * [`master`](https://github.com/swashdev/SwashRL/tree/master)  
   This is the central branch for the project and is meant to contain the most
   recent stable or "release" versions of the code and the most up-to-date
   documentation.  Only merge to `master` if code is stable and ready for
   release.  Tagged commits will always come from `master`.
   [`hotfix-*` branches](#Production-Branches) should branch from `master`.

 * [`dev`](https://github.com/swashdev/SwashRL/tree/dev)  
   This is the branch used when modifying code files.  Merge this branch to
   `master` with `--no-ff` when the code is ready for release.  When working
   on a new feature, create a [`feature-*` branch](#Production-Branches) off
   of `dev` and merge back when the feature is complete.

 * [`spring-cleaning`](https://github.com/swashdev/SwashRL/tree/spring-cleaning)  
   This is a special branch originally created towards the end of February
   2020 to facilitate a radical cleanup and restructuring of SwashRL's code
   base and make it compliant with an updated style guide.  It is essentially
   equivalent to a `feature-*` branch in scope, except that it is not working
   on any particular feature and is publically available for peer review.  Any
   time a "spring cleaning" project is done, it will be merged into `dev`.

### Production Branches

These are private branches used when working on individual copies of the
source code, except that some of them may be pushed to a public repo
temporarily in order to collaborate on preparation for an upcoming release or
if assistance is needed.  As above, the branches are listed according to how
close they are to the `master` branch.

 * `hotfix-*`
   These branches are used to quickly fix bugs in `master` release versions of
   SwashRL.  These branches should be very short-lived; larger bugfixes will
   have to happen in `dev`.  Name the branch according to what the version
   number will be after the hotfix is complete, e.g. `hotfix-0.032`.  Do not
   merge `hotfix-*` branches to `master` until they are ready for release.

 * `feature-*`  
   Whenever you are working on adding a new feature, merge `dev` into a
   feature branch in a private repository.  The naming convention should give
   away what its intended feature is; for example, `feature-colors` was used
   for an overhaul of the color storage system in SwashRL's display code.  All
   of the work done in this branch should go toward working on that particular
   feature, not bug fixes in the existing source code.  Merge back to `dev`
   with `--no-ff` when the feature is complete and then delete the branch.

### Special Branches

These branches are public, but are not related an a fundamental level to the
production of source code.

 * [`gh-pages`](https://github.com/swashdev/SwashRL/tree/gh-pages)  
   This is an orphan branch which is used to host SwashRL's GitHub Pages site.
   Obviously, you shouldn't be pushing any code files here.  If you are not
   on GitHub, you might not see this branch, because it's only relevant to
   GitHub Pages.

 * [`branch-readme`](https://github.com/swashdev/SwashRL/tree/branch-readme)  
   This is an orphan branch which is used to hold this readme file.  This
   branch should only ever be updated in order to update this readme file.
   **No code should be pushed to this branch as it may cause a licensing issue
   due to the lack of license notices in this branch**.


history.md
---
layout: default
title: History
description: A brief history of the SwashRL project, from conception to present day.
permalink: /history/
---

Table of Contents
-----------------

* [Prehistory]
* [Slumbering Dragon]
* [FORRNIF]
* [Spelunk!]
  * [The First Public git Repository]
  * [The Translation into D]
* [SwashRL]
  * [SDL Support]

[Jump to the current version](#0035)

[Back to top]: #table-of-contents

[Prehistory]: #prehistory
[Slumbering Dragon]: #slumbering-dragon
[FORRNIF]: #forrnif
[Spelunk!]: #spelunk
[SwashRL]: #swashrl

[The First Public git Repository]: #0019
[The Translation into D]: #0022
[SDL Support]: #0024

[0.map]: #0map
[0.0.0]: #0map
[0.move]: #0move
[0.0.1]: #0move
[0.colission]: #0colission
[0.0.2]: #0colission
[0.hazards]: #0hazards
[0.0.3]: #0hazards
[0.text-effects]: #0text-effects
[0.0.4]: #0text-effects
[0.monster-display]: #0monster-display
[0.0.5]: #0monster-display
[0.monster-colission]: #0monster-colission
[0.0.6]: #0monster-colission
[0.monster-fight]: #0monster-fight
[0.0.7]: #0monster-fight
[0.monster-move]: #0monster-move
[0.0.8]: #0monster-move
[0.monster-move-chase]: #0monster-move-chase
[0.0.9]: #0monster-move-chase
[0.monster-move-fly]: #0monster-move-fly
[0.0.10]: #0monster-move-fly
[0.0.11]: #0011
[0.0.12]: #0012
[0.0.13]: #0013
[0.0.14]: #0014
[0.0.15]: #0015
[0.0.16]: #0016
[0.017]: #0017
[0.018]: #0018
[0.019]: #0019
[0.020]: #0020
[0.020-1]: #0020-1
[0.021]: #0021
[0.022]: #0022
[0.023]: #0023
[0.024]: #0024
[0.025]: #0025
[0.026]: #0026
[0.027]: #0027
[0.028]: #0028
[0.029]: #0029
[0.029-1]: #0029-1
[0.030]: #0030
[0.030-1]: #0030-1
[0.031]: #0031
[0.031 Christmas Update]: #0031-christmas-update
[0.031-christmas-update]: #0031-christmas-update
[0.032]: #0032
[0.033]: #0033
[0.034]: #0034
[0.035]: #0035

Prehistory
----------

SwashRL is actually the modern reimagining of a game I started writing in
C# way back in 2005, while I was still in high school.  At that time it was
simply named Crawl, and served as my homage to '90s text adventure games and
early dungeon crawlers.  At the time I didn't even know that Roguelikes were a
thing, but it ended up being very similar to rudimentary Roguelikes.  I came
back to this project many times, porting it first to C++ and then Python.
After a while writing a Roguelike became the de facto way I learned new
programming languages.

Despite the name, Crawl had nothing to do with Linley's Dungeon Crawl or its
derivative Stone Soup, which is the reason I ended up changing the name.  In
fact in its original form it hardly resembled a Roguelike at all.  The game
generated a dungeon by randomly generating two walls stretching from left to
right.  The player (`P`) was required to proceed from left to right down
twenty levels, fighting monsters (`O`) with randomly-generated names and
collecting treasure (`X`) which included weapons and armor.  The dungeon also
had stalactites (`V`) and test tubes (`U`) which acted as "obstacles."

It's safe to say that Crawl was not a very well-designed game, but it remained
close to my heart as one of my first really ambitious programming projects.

Slumbering Dragon
-----------------

In 2015, I was experimenting with Linux virtual machines and messing around
with various utilities when I decided to try writing a simple game to see if I
could work out how to get curses working in C.  I called this game Slumbering
Dragon, or simply SD.  Because I wasn't planning on releasing it, I didn't
bother with proper version numbers (or checking to see if the name was already
taken), which is why in historical versions of the code you instead see a 0
followed by a short one- or two-word phrase describing what that version had
introduced.

### 0.map

2015-08-24 12:31 EST

First version of the current source code.  Displayed a test map.
Apart from the ability to quit the program, there was no user interaction.
The test map has changed over time as new game features has been implemented,
but its general layout has remained largely the same.

[Back to top]

### 0.move

2015-08-24 14:10 EST

Movement implemented: The player character is able to move around with the
traditional HJKLYUBN movement keys.  Due to a lack of collision detection, it
is possible to move outside the boundaries of the map and cause the game to
crash due to an array index exception.

[Back to top]

### 0.colission

2015-08-24 20:14 EST

Collision detection implemented: Walking into a wall produces the message
"Ouch!  You walk straight into a wall!".  The typo in the version name is
accurate to the actual label I gave this version of the program.

[Back to top]

### 0.hazards

2015-08-24 21:29 EST

The first hazard is implemented: If the player steps in the water, they die.

[Back to top]

### 0.text-effects

2015-08-29 17:58 EST

"Text effects" (inverted colors for walls) implemented.  I'm not certain, but
I believe this is also the update where I started using the "dim" effect to
draw the player character in a darker color after they die, which still exists
in the curses interface at time of writing.

[Back to top]

### 0.monster-display

2015-08-30 14:44 EST

The first monster (a goobling) is implemented.  It is displayed on the map but
does not interact with anything.

[Back to top]

### 0.monster-colission

2015-08-30 19:56 EST

The player now collides with monsters, producing the message "You bump into a
*%s*.  It doesn't look too pleased."

[Back to top]

### 0.monster-fight

2015-09-02 10:23 EST

Rudimentary combat is implemented:  If the player runs in to the goobling, the
player attacks and then the goobling attacks back.  The goobling has no AI and
therefore can not initiate combat; all interactions originate from the player
at this stage.

This was the first version of Slumbering Dragon to have Windows support; if
the `WINDOWS` flag was set, the program would compile with pdcurses rather
than ncurses.  

[Back to top]

### 0.monster-move

2015-09-03 21:43 EST


src/map.c
/*
 * Copyright (c) 2017 Philip Pavlick.  All rights reserved.
 *
 * Spelunk! may be modified and distributed, but comes with NO WARRANTY!
 * See license.txt for details.
 */
#include "global.h"

map new_map()
{
  map nu;
  nu.m_siz = 0;
  return nu;
}

void add_mon( map* mp, monst mn )
{
  unsigned int mndex = mp->m_siz;
  if( mndex < NUMTILES )
  {
    mp->m[mndex] = mn;
    mp->m_siz = mndex + 1;
  }
}

void remove_mon( map* mp, uint16 index )
{
  /* To remove a monster in a map's mon array, move all monsters that are
   * past it in the array up, thus overwriting it. */
  if( index < mp->m_siz )
  {
    int mn, max;
    for( mn = index + 1, max = mp->m_siz; mn < max; mn++ )
    {
      mp->m[mn - 1] = mp->m[mn];
    }
    mp->m_siz = max - 1;
  }
}

map test_map()
{
  map nu = new_map();

  uint8 y, x;
  for( y = 0; y < MAP_Y; y++ )
    for( x = 0; x < MAP_X; x++ )
    {
      nu.i[y][x] = No_item;
      if( y == 0 || y == MAP_y || x == 0 || x == MAP_x )
      {
	nu.t[y][x] = T_WALL;
      }
      else
      {
	if( (y < 13 && y > 9) && ((x > 19 && x < 24) || (x < 61 && x > 56)) )
	  nu.t[y][x] = T_WALL;
	else
	{
	  if( (y < 13 && y > 9) && (x > 30 && x < 50) )
	    nu.t[y][x] = T_WATER;
	  else
	    nu.t[y][x] = T_FLOOR;
	}
      }
    }

  monst goobling = new_monst_at( 'g', "goobling", 0, 0, 2, 2, 0, 10, 2, 0, 2,
                                 1000, 60, 20 );

  add_mon( &nu, goobling );

#if 0
  goobling.x = 50;
  add_mon( &nu, goobling );
  goobling.y = 10;
  add_mon( &nu, goobling );
#endif

  /* test items */

  /* a test item "old sword" which grants a +2 bonus to the player's
   * attack roll */
  item old_sword = { .sym = symdata( '(', A_NORMAL ),
                     .name = "old sword",
                     .type = ITEM_WEAPON, .equip = EQUIP_NO_ARMOR,
                     .addd = 0, .addm = 2 };
  nu.i[10][5] = old_sword;

  item ring = { .sym = symdata( '=', A_NORMAL ),
                .name = "tungsten ring",
                .type = ITEM_JEWELERY, .equip = EQUIP_JEWELERY_RING,
                .addd = 0, .addm = 0 };
  nu.i[10][2] = ring;
  nu.i[10][1] = ring;
  item helmet = { .sym = symdata( ']', A_NORMAL ),
                  .name = "hat",
                  .type = ITEM_ARMOR, .equip = EQUIP_HELMET,
                  .addd = 0, .addm = 0 };
  nu.i[10][3] = helmet;
  item scarf = { .sym = symdata( ']', A_NORMAL ),
                 .name = "fluffy scarf",
                 .type = ITEM_ARMOR, .equip = EQUIP_JEWELERY_NECK,
                 .addd = 0, .addm = 0 };
  nu.i[11][3] = scarf;
  item tunic = { .sym = symdata( ']', A_NORMAL ),
                 .name = "tunic",
                 .type = ITEM_ARMOR, .equip = EQUIP_CUIRASS,
                 .addd = 0, .addm = 0 };
  nu.i[12][3] = tunic;
  item gloves = { .sym = symdata( ']', A_NORMAL ),
                  .name = "pair of leather gloves",
                  .type = ITEM_ARMOR, .equip = EQUIP_BRACERS,
                  .addd = 0, .addm = 1 };
  nu.i[13][3] = gloves;
  item pants = { .sym = symdata( ']', A_NORMAL ),
                 .name = "pair of trousers",
                 .type = ITEM_ARMOR, .equip = EQUIP_GREAVES,
                 .addd = 0, .addm = 0 };
  nu.i[14][3] = pants;
  item kilt = { .sym = symdata( ']', A_NORMAL ),
                .name = "plaid kilt",
                .type = ITEM_ARMOR, .equip = EQUIP_KILT,
                .addd = 0, .addm = 0 };
  nu.i[15][3] = kilt;
  item boots = { .sym = symdata( ']', A_NORMAL ),
                 .name = "pair of shoes",
                 .type = ITEM_ARMOR, .equip = EQUIP_FEET,
                 .addd = 0, .addm = 0 };
  nu.i[16][3] = boots;
  item tailsheath = { .sym = symdata( ']', A_NORMAL ),
                      .name = "leather tailsheath",
                      .type = ITEM_ARMOR, .equip = EQUIP_TAIL,
                      .addd = 0, .addm = 1 };
  nu.i[17][3] = tailsheath;
  
  return nu;
}

/*
################################################################################
#..............................................................................#
#..............................................................................#
#..............................................................................#
#..............................................................................#
#..............................................................................#
#..............................................................................#
#..............................................................................#
#..............................................................................#
#..............................................................................#
#...................###..................................###...................#
#...................###..................................###...................#
#...................###..................................###...................#
#..............................................................................#
#..............................................................................#
#..............................................................................#
#..............................................................................#
#..............................................................................#
#..............................................................................#
#..............................................................................#
#..............................................................................#
#..............................................................................#
#..............................................................................#
################################################################################
*/
